server/src/main/java/org/elasticsearch/action/admin/indices/rollover/AutoShardCondition.java

```diff
@@ -0,0 +1,138 @@
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Condition for automatically increasing the number of shards for a data stream. The value is computed when the condition is
+ * evaluated.
+ */
+public class AutoShardCondition extends Condition<String> {
+    public static final String NAME = "auto_sharding";
+
+    public static final ParseField AUTO_SHARDING_TYPE = new ParseField("type");
+    public static final ParseField CURRENT_NUMBER_OF_SHARDS = new ParseField("current_number_of_shards");
+    public static final ParseField TARGET_NUMBER_OF_SHARDS = new ParseField("target_number_of_shards");
+    public static final ParseField WRITE_LOAD = new ParseField("write_load");
+    private final AutoShardingResult autoShardingResult;
+
+    public AutoShardCondition(AutoShardingResult autoShardingResult) {
+        super(NAME, Type.AUTOMATIC);
+        if (autoShardingResult.type().equals(AutoShardingType.INCREASE_SHARDS) == false
+            && autoShardingResult.type().equals(AutoShardingType.DECREASE_SHARDS) == false) {
+            throw new IllegalArgumentException("The autosharding condition only supports the INCREASE and DECREASE shards types");
+        }
+        this.value = toCSVRep(autoShardingResult);
+        this.autoShardingResult = autoShardingResult;
+    }
+
+    static String toCSVRep(AutoShardingResult autoShardingResult) {
+        return String.join(
+            ",",
+            List.of(
+                AUTO_SHARDING_TYPE.getPreferredName() + "=" + autoShardingResult.type(),
+                CURRENT_NUMBER_OF_SHARDS.getPreferredName() + "=" + autoShardingResult.currentNumberOfShards(),
+                TARGET_NUMBER_OF_SHARDS.getPreferredName() + "=" + autoShardingResult.targetNumberOfShards(),
+                WRITE_LOAD.getPreferredName() + "=" + autoShardingResult.writeLoad()
+            )
+        );
+    }
+
+    // visible for testing
+    public static AutoShardingResult fromCSVRep(String csvRep) {
+        if (Strings.isNullOrBlank(csvRep)) {
+            throw new IllegalArgumentException("The provided csv representation cannot be null");
+        }
+        String[] fieldsAndValues = csvRep.split(",");
+        AutoShardingType type = null;
+        Integer currentNumberOfShards = null;
+        Integer targetNumberOfShards = null;
+        Double writeLoad = null;
+        for (String fieldsAndValue : fieldsAndValues) {
+            String[] keyValue = fieldsAndValue.split("=");
+            if (keyValue.length != 2) {
+                throw new IllegalArgumentException("Invalid auto sharding condition representaion: " + csvRep);
```

Comment:
I'm worried that in the future if this were to break and throw exceptions, it would wedge a node from starting up or something, or else spam the logs forever and ever. I sent you an alternative idea, what do you think of that?